src/mrpro/algorithms/_lbfgs.py

```diff
@@ -0,0 +1,86 @@
+"""LBFGS for solving non-linear minimization problems."""
+
+# Copyright 2023 Physikalisch-Technische Bundesanstalt
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#       http://www.apache.org/licenses/LICENSE-2.0
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+from torch.optim import LBFGS
+
+
+def lbfgs(
+    f,
+    params,
+    lr=1,
+    max_iter=100,
+    max_eval=100,
+    tolerance_grad=1e-07,
+    tolerance_change=1e-09,
+    history_size=10,
+    line_search_fn='strong_wolfe',
+):
+    """LBFGS for non-linear minimization problems.
+
+    Parameters
+    ----------
+    f
+        scalar function to be minimized
+    params
+        list with parameters containing the initial guess of the solution
+    lr, optional
+        learning rate
+    max_iter, optional
+        maximal number of iterations, by default 100
+    max_eval, optional
+        maximal number of evaluations of f per optimization step,
+        by default 100
+    tolerance_grad, optional
+        termination tolerance on first order optimality,
+        by default 1e-07
+    tolerance_change, optional
+        termination tolerance on function value/parameter changes, by default 1e-09
+    history_size, optional
+        update history size, by default 10
+    line_search_fn, optional
+        line search algorithm, either ‘strong_wolfe’ or None,
+        by default "strong_wolfe"
+
+    Returns
+    -------
+        optimized parameter p
+    """
+
+    # enable gradient calculation
+    for p in params:
```

Comment:
detach().clone().requiere_grad_() to create new tensors 
enabling require grad on the inputs is surprising and will break.

Reply:
or its the job of the calling function (which will also be part of mrpro)

so either way, I would recommend not having this loop here.